File: algorithms/cpp/numberofBoomerangs/numberofBoomerangs.cpp
// Source: https://leetcode.com/problems/number-of-boomerangs/
// Author: Miao Zhang
// Date:   2021-02-12

class Solution {
public:
    int numberOfBoomerangs(vector<vector<int>>& points) {
        int res = 0;
        for (int i = 0; i < points.size(); i++) {
            unordered_map<int, int> dist(points.size());
            for (int j = 0; j < points.size(); j++) {
                int dx = points[i][0] - points[j][0];
                int dy = points[i][1] - points[j][1];
                dist[dx * dx + dy * dy]++;
            }
            for (auto &d: dist) {
                res += d.second * (d.second - 1);
            }
        }
        return res;
    }
};

File: algorithms/python/numberofBoomerangs/numberofBoomerangs.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/number-of-boomerangs/
# Author: Miao Zhang
# Date:   2021-02-12

class Solution:
    def numberOfBoomerangs(self, points: List[List[int]]) -> int:
        def dis(p1: List[int], p2: List[int]) -> int:
            return (p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2
        
        res = 0
        for i in points:
            record = {}
            for j in points:
                if i != j:
                    distance = dis(i, j)
                    record[distance] = record.get(distance, 0) + 1
            for val in record.values():
                if val >= 2:
                    res += (val - 1) * val
        return res
